ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/manager/EthPeersTest.java

```diff
@@ -64,4 +86,187 @@ public void comparesPeersWithTdAndNoHeight() {
+    assertRequestFailure(pendingRequest, NoAvailablePeersException.class);
+  }
+
+  @Test
+  public void shouldFailWithPeerNotConnectedIfPeerRequestThrows() throws Exception {
+    final RespondingEthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
+    when(peerRequest.sendRequest(peer.getEthPeer())).thenThrow(new PeerNotConnected("Oh dear"));
+    final PendingPeerRequest pendingRequest =
+        ethPeers.executePeerRequest(peerRequest, 100, Optional.empty());
+
+    assertRequestFailure(pendingRequest, PeerDisconnectedException.class);
+  }
+
+  @Test
+  public void shouldDelayExecutionUntilPeerHasCapacity() throws Exception {
+    final RespondingEthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
+    useAllAvailableCapacity(peer.getEthPeer());
+
+    final PendingPeerRequest pendingRequest =
+        ethPeers.executePeerRequest(peerRequest, 100, Optional.empty());
+    verifyZeroInteractions(peerRequest);
+
+    freeUpCapacity(peer.getEthPeer());
+
+    verify(peerRequest).sendRequest(peer.getEthPeer());
+    assertRequestSuccessful(pendingRequest);
+  }
+
+  @Test
+  public void shouldDelayExecutionUntilPeerWithSufficientHeightHasCapacity() throws Exception {
+    // Create a peer that has available capacity but not the required height
+    EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 10);
+
+    final RespondingEthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
+    useAllAvailableCapacity(peer.getEthPeer());
+
+    final PendingPeerRequest pendingRequest =
+        ethPeers.executePeerRequest(peerRequest, 100, Optional.empty());
+    verifyZeroInteractions(peerRequest);
+
+    freeUpCapacity(peer.getEthPeer());
+
+    verify(peerRequest).sendRequest(peer.getEthPeer());
+    assertRequestSuccessful(pendingRequest);
+  }
+
+  @Test
+  public void shouldNotExecuteAbortedRequest() throws Exception {
+    final RespondingEthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
+    useAllAvailableCapacity(peer.getEthPeer());
+
+    final PendingPeerRequest pendingRequest =
+        ethPeers.executePeerRequest(peerRequest, 100, Optional.empty());
+    verifyZeroInteractions(peerRequest);
+
+    pendingRequest.abort();
+
+    freeUpCapacity(peer.getEthPeer());
+
+    verifyZeroInteractions(peerRequest);
+    assertRequestFailure(pendingRequest, CancellationException.class);
+  }
+
+  private void freeUpCapacity(final EthPeer ethPeer) {
+    ethPeers.dispatchMessage(ethPeer, new EthMessage(ethPeer, NodeDataMessage.create(emptyList())));
+  }
+
+  private void useAllAvailableCapacity(final EthPeer peer) throws PeerNotConnected {
+    while (peer.hasAvailableRequestCapacity()) {
+      useRequestSlot(peer);
+    }
+    assertThat(peer.hasAvailableRequestCapacity()).isFalse();
+  }
+
+  private void useRequestSlot(final EthPeer peer) throws PeerNotConnected {
+    peer.getNodeData(emptyList());
```

Comment:
Seems like `getNodeData()` should throw an exception here.  For now, it's probably worth at least sending a non-empty list to this method in order to future-proof these tests.

Reply:
Done.